File: medium/e75/typescript/solution_test.ts
import { sortColors } from "./solution"

function testSortColors(): string {
    let input1: number[][] = [[2, 0, 2, 1, 1, 0], [2, 0, 1]];

    let expectations: number[][] = [[0, 0, 1, 1, 2, 2], [0, 1, 2]];

    if (input1.length != expectations.length) {
        return "test is invalid";
    }

    for (let i = 0; i < expectations.length; i++) {
        const origInput: number[] = Object.assign([], input1[i])

        sortColors(input1[i]);

        let msg = `bad result for ${origInput}: got ${input1[i]} expected ${expectations[i]}`;

        if (origInput.length != input1[i].length) {
            return msg;
        }

        for (let j: number = 0; j < expectations[i].length; j++) {
            if (expectations[i][j] != input1[i][j]) {
                return msg;
            }
        }
    }

    return "PASS";
}

console.log(testSortColors());

File: medium/e451/cpp/solution_test.cpp
#include "solution.h"

#include <cassert>
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

char asciitolower(char in) {
    if (in <= 'Z' && in >= 'A')
        return in - ('Z' - 'z');
    return in;
}

int main() {
    std::vector<std::string> input1 = { "tree", "cccaaa", "Aabb" };

    std::vector<std::string> expectations = { "eert" , "aaaccc", "bbAa" };

    if (input1.size() != expectations.size()) {
        std::cerr << "test is invalid";
        return 0;
    }

    Solution s;
    for (int i = 0; i < input1.size(); i++) {
        assert(1 <= input1[i].size() && input1[i].size() <= 5 * pow(10, 5));

        std::string result = s.frequencySort(input1[i]);

        std::string lowercasedResult = result;
        std::transform(lowercasedResult.begin(), lowercasedResult.end(), lowercasedResult.begin(), asciitolower);
        std::string lowercasedExpectation = expectations[i];
        std::transform(lowercasedExpectation.begin(), lowercasedExpectation.end(), lowercasedExpectation.begin(), asciitolower);

        if (lowercasedResult != lowercasedExpectation) {
            std::cerr << "FAIL: bad result for \"" << input1[i];
            std::cerr << "\": got \"" << result;
            std::cerr << "\" expected \"" << expectations[i];
            std::cerr << "\"" << std::endl;

            return 0;
        }
    }

    std::cout << "PASS";

    return 0;
}